File: reference/python/hbprep/segment.py
# -*- coding: utf-8 -*-
"""Satzsegmentierung (DE/EN), regelbasiert, ohne Fremdbibliotheken.

Kernproblem beim Hörbuch-Skript: der Satz ist die Einheit, die der Sprecher
atmet.  Die Segmentierung muss deshalb vor allem *keine* falschen Grenzen
setzen – bei »…«, sagte er.  ist das ein Satz, nicht zwei.
"""

from __future__ import annotations

import re

# Abkürzungen, nach denen ein Punkt keine Satzgrenze ist.
ABBREV = {
    # deutsch
    "z", "b", "zb", "dh", "ua", "ggf", "bzw", "ca", "evtl", "usw", "usf", "vgl",
    "bspw", "inkl", "exkl", "max", "min", "mind", "sog", "u", "a", "d", "h",
    "nr", "abb", "tab", "kap", "bd", "jh", "jhd", "chr", "geb", "gest", "verst",
    "ebd", "ff", "f", "s", "str", "pl", "hrsg", "aufl", "od", "bzgl", "etc",
    "dr", "prof", "dipl", "ing", "med", "phil", "jur", "rer", "nat", "hc",
    "hr", "fr", "frl", "mr", "mrs", "ms", "st", "hl", "kl", "gr", "nachm",
    "vorm", "mio", "mrd", "tsd", "gez", "i", "v", "m", "e", "o", "n", "z.b",
    # englisch
    "mt", "jr", "sr", "vs", "eg", "ie", "approx", "dept", "univ", "co", "inc",
    "ltd", "cf", "al", "ed", "vol", "pp", "fig", "no", "esp",
}
MONTHS = {
    "januar", "februar", "märz", "april", "mai", "juni", "juli", "august",
    "september", "oktober", "november", "dezember", "jänner",
    "january", "february", "march", "may", "june", "july", "october", "december",
}

OPENERS = "«»„“”‘’‚‹›\"'([{–—-"
CLOSERS = "«»“”‘’‹›\"')]}"

_END = re.compile(r"[.!?…]+")
_UPPER = re.compile(r"[A-ZÄÖÜÁÀÂÉÈÊÍÎÓÔÚÛÑÇŠŽŁ0-9]")
_WORD = re.compile(r"[^\W\d_]+", re.UNICODE)


def _prev_token(text: str, i: int) -> str:
    j = i
    while j > 0 and (text[j - 1].isalnum() or text[j - 1] in "ÄÖÜäöüß."):
        j -= 1
    return text[j:i]


def sentence_spans(text: str) -> list[tuple[int, int]]:
    """Liefert (start, ende)-Offsets aller Sätze eines Blocks."""
    spans, start, n = [], 0, len(text)
    for m in _END.finditer(text):
        i, j = m.start(), m.end()
        punct = m.group()

        # Trailing Anführungs-/Klammerzeichen mit einschließen
        k = j
        while k < n and text[k] in CLOSERS:
            k += 1

        rest = text[k:]
        nxt = rest.lstrip()
        # Satzende nur, wenn danach Ende oder Großbuchstabe/Anführung/Gedankenstrich
        if nxt and not (_UPPER.match(nxt[0]) or nxt[0] in OPENERS):
            continue
        if nxt and rest and not rest[0].isspace() and k < n:
            # z.B. "Wort.Wort" – meist Fehler in der Quelle, trotzdem trennen,
            # aber nicht mitten in Ziffern (3.14) oder Domains
            if text[i - 1:i].isdigit() and nxt[0].isdigit():
                continue

        if punct == ".":
            tok = _prev_token(text, i).strip(".").lower()
            if tok in ABBREV and len(tok) <= 5:
                continue
            if len(tok) == 1 and tok.isalpha():          # Initialen: J. R. R.
                continue
            if tok.isdigit():                            # Ordinalzahl / Datum
                w = _WORD.search(nxt or "")
                if not nxt:
                    pass
                elif nxt[0].islower() or (w and w.group().lower() in MONTHS):
                    continue
                elif len(tok) <= 2 and int(tok) <= 31:   # "am 3. Oktober", "1. Kapitel"
                    continue

        if punct.startswith("…") or punct == "...":
            if not nxt or (nxt and not (_UPPER.match(nxt[0]) or nxt[0] in "»„“\"")):
                continue

        end = k
        if text[start:end].strip():
            spans.append((start, end))
        start = k
        while start < n and text[start].isspace():
            start += 1

    if start < n and text[start:].strip():
        spans.append((start, n))
    return spans


def word_count(text: str) -> int:
    return len(_WORD.findall(text))


def segment_doc(doc: dict) -> dict:
    """Ergänzt jeden Block um 'sent': Liste von [start, end]."""
    idx = 0
    for ch in doc["chapters"]:
        ch_idx = 0
        for b in ch["blocks"]:
            if b["type"] in ("h1", "h2"):
                # Überschriften bekommen keine Satznummer: die Nummerierung soll
                # zwischen Report, overrides.json und Reader identisch sein.
                b["sent"], b["sid"] = [], []
                b["words"] = word_count(b["text"])
                continue
            spans = sentence_spans(b["text"])
            b["sent"] = [[s, e] for s, e in spans]
            b["sid"] = []
            for _ in spans:
                idx += 1
                ch_idx += 1
                b["sid"].append(ch_idx)
            b["words"] = word_count(b["text"])
        ch["n_sentences"] = ch_idx
    doc["meta"]["n_sentences"] = idx
    doc["meta"]["n_words"] = sum(b["words"] for c in doc["chapters"] for b in c["blocks"])
    return doc

File: packages/core/src/pipeline/segment.ts
/**
 * Satzsegmentierung (DE/EN), regelbasiert.
 *
 * Der Satz ist die Einheit, die der Sprecher atmet. Die Segmentierung muss
 * deshalb vor allem *keine* falschen Grenzen setzen: »…«, sagte er. ist ein
 * Satz, nicht zwei – nach dem schließenden Anführungszeichen folgt ein Komma.
 */
import { LETTER, isAlnum, isLower, isSpace, wordCount } from "../text.js";
import type { Doc } from "./types.js";

/** Abkürzungen, nach denen ein Punkt keine Satzgrenze ist. */
export const ABBREV = new Set([
  // deutsch
  "z", "b", "zb", "dh", "ua", "ggf", "bzw", "ca", "evtl", "usw", "usf", "vgl",
  "bspw", "inkl", "exkl", "max", "min", "mind", "sog", "u", "a", "d", "h",
  "nr", "abb", "tab", "kap", "bd", "jh", "jhd", "chr", "geb", "gest", "verst",
  "ebd", "ff", "f", "s", "str", "pl", "hrsg", "aufl", "od", "bzgl", "etc",
  "dr", "prof", "dipl", "ing", "med", "phil", "jur", "rer", "nat", "hc",
  "hr", "fr", "frl", "mr", "mrs", "ms", "st", "hl", "kl", "gr", "nachm",
  "vorm", "mio", "mrd", "tsd", "gez", "i", "v", "m", "e", "o", "n", "z.b",
  // englisch
  "mt", "jr", "sr", "vs", "eg", "ie", "approx", "dept", "univ", "co", "inc",
  "ltd", "cf", "al", "ed", "vol", "pp", "fig", "no", "esp",
]);

const MONTHS = new Set([
  "januar", "februar", "märz", "april", "mai", "juni", "juli", "august",
  "september", "oktober", "november", "dezember", "jänner",
  "january", "february", "march", "may", "june", "july", "october", "december",
]);

export const OPENERS = "«»„“”‘’‚‹›\"'([{–—-";
export const CLOSERS = "«»“”‘’‹›\"')]}";

const RE_END = /[.!?…]+/gu;
const RE_UPPER = /^[A-ZÄÖÜÁÀÂÉÈÊÍÎÓÔÚÛÑÇŠŽŁ0-9]/u;
const RE_FIRST_WORD = new RegExp(`${LETTER}+`, "u");

function prevToken(text: string, i: number): string {
  let j = i;
  while (j > 0 && (isAlnum(text[j - 1]) || "ÄÖÜäöüß.".includes(text[j - 1]!))) j--;
  return text.slice(j, i);
}

/** Satzgrenzen eines Blocks als [start, ende)-Offsets. */
export function sentenceSpans(text: string): [number, number][] {
  const spans: [number, number][] = [];
  const n = text.length;
  let start = 0;
  RE_END.lastIndex = 0;
  for (let m = RE_END.exec(text); m; m = RE_END.exec(text)) {
    const i = m.index;
    const j = i + m[0].length;
    const punct = m[0];

    let k = j;
    while (k < n && CLOSERS.includes(text[k]!)) k++;

    const rest = text.slice(k);
    const nxt = rest.trimStart();
    // Grenze nur vor Textende, Großbuchstabe, Ziffer oder öffnendem Zeichen
    if (nxt && !(RE_UPPER.test(nxt) || OPENERS.includes(nxt[0]!))) continue;
    if (nxt && rest && !isSpace(rest[0]) && k < n) {
      if (/[0-9]/.test(text[i - 1] ?? "") && /[0-9]/.test(nxt[0]!)) continue;
    }

    if (punct === ".") {
      const tok = prevToken(text, i).replace(/^\.+|\.+$/g, "").toLowerCase();
      if (ABBREV.has(tok) && tok.length <= 5) continue;
      if (tok.length === 1 && /^\p{L}$/u.test(tok)) continue; // Initialen: J. R. R.
      if (/^[0-9]+$/.test(tok) && nxt) {
        // Ordinalzahl / Datum: »am 3. Oktober«, »1. Kapitel«
        const w = RE_FIRST_WORD.exec(nxt);
        if (isLower(nxt[0]) || (w && MONTHS.has(w[0].toLowerCase()))) continue;
        if (tok.length <= 2 && Number(tok) <= 31) continue;
      }
    }

    if (punct.startsWith("…") || punct === "...") {
      if (!nxt || !(RE_UPPER.test(nxt) || "»„“\"".includes(nxt[0]!))) continue;
    }

    if (text.slice(start, k).trim()) spans.push([start, k]);
    start = k;
    while (start < n && isSpace(text[start])) start++;
  }
  if (start < n && text.slice(start).trim()) spans.push([start, n]);
  return spans;
}

/** Ergänzt Blöcke um Satzgrenzen, Satznummern und Wortzahlen. */
export function segmentDoc(doc: Doc): Doc {
  let idx = 0;
  for (const ch of doc.chapters) {
    let chIdx = 0;
    for (const b of ch.blocks) {
      b.words = wordCount(b.text);
      if (b.type === "h1" || b.type === "h2") {
        // Überschriften bekommen keine Satznummer.
        b.sent = [];
        b.sid = [];
        continue;
      }
      b.sent = sentenceSpans(b.text);
      b.sid = b.sent.map(() => {
        idx++;
        return ++chIdx;
      });
    }
    ch.n_sentences = chIdx;
  }
  doc.meta.n_sentences = idx;
  doc.meta.n_words = doc.chapters.reduce((a, c) => a + c.blocks.reduce((x, b) => x + (b.words ?? 0), 0), 0);
  return doc;
}
